src/probability/ident_distrib.lean

```diff
@@ -0,0 +1,277 @@
+in this file.
+
+## Main definitions and results
+
+* `ident_distrib f g μ ν` registers that the image of `μ` under `f` coincides with the image of `ν`
+  under `g` (and that `f` and `g` are almost everywhere measurable, as otherwise the image measures
+  don't make sense). The measures can be kept implicit as in `ident_distrib f g` if the spaces
+  are registered as measure spaces.
+* `ident_distrib.comp`: being identically distributed is stable under composition with measurable
+  maps.
+
+There are two main kind of lemmas, under the assumption that `f` and `g` are identically
+distributed: lemmas saying that two quantities computed for `f` and `g` are the same, and lemmas
+saying that if `f` has some property then `g` also has it. The first kind is registered as
+`ident_distrib.foo_eq`, the second one as `ident_distrib.foo_snd` (in the latter case, to deduce
+a property of `f` from one of `g`, use `h.symm.foo_snd` where `h : ident_distrib f g μ ν`). For
+instance:
+
+* `ident_distrib.measure_mem_eq`: if `f` and `g` are identically distributed, then the probabilities
+  that they belong to a given measurable set are the same.
+* `ident_distrib.integral_eq`: if `f` and `g` are identically distributed, then their integrals
+  are the same.
+* `ident_distrib.variance_eq`: if `f` and `g` are identically distributed, then their variances
+  are the same.
+
+* `ident_distrib.ae_strongly_measurable_snd`: if `f` and `g` are identically distributed and `f`
+  is almost everywhere strongly measurable, then so is `g`.
+* `ident_distrib.mem_ℒp_snd`: if `f` and `g` are identically distributed and `f`
+  belongs to `ℒp`, then so does `g`.
+
+We also register several dot notation shortcuts for convenience.
+For instance, if `h : ident_distrib f g μ ν`, then `h.sq` states that `f^2` and `g^2` are
+identically distributed, and `h.norm` states that `∥f∥` and `∥g∥` are identically distributed, and
+so on.
+-/
+
+open measure_theory filter finset
+
+noncomputable theory
+
+open_locale topological_space big_operators measure_theory ennreal nnreal
+
+variables {α β γ δ : Type*} [measurable_space α] [measurable_space β]
+  [measurable_space γ] [measurable_space δ]
+
+namespace probability_theory
+
+/-- Two functions defined on two (possibly different) measure spaces are identically distributed if
+their image measures coincide. This only makes sense when the functions are ae measurable
+(as otherwise the image measures are not defined), so we require this as well in the definition. -/
+structure ident_distrib
+  (f : α → γ) (g : β → γ) (μ : measure α . volume_tac)  (ν : measure β . volume_tac) : Prop :=
+(ae_measurable_fst : ae_measurable f μ)
+(ae_measurable_snd : ae_measurable g ν)
+(map_eq : measure.map f μ = measure.map g ν)
+
+namespace ident_distrib
+
+open topological_space
+
+variables {μ : measure α} {ν : measure β} {f : α → γ} {g : β → γ}
+
+protected lemma ae_measurable (h : ident_distrib f g μ ν) :
+  ae_measurable f μ := h.ae_measurable_fst
```

Comment:
To me, that lemma looks like `ae_measurable_fst` but with the drawback of an ambiguous name, and I don't see why it is desirable. Why do you want this one?

Reply:
I have a naming problem with `ae_strongly_measurable`, as per my naming conventions `ae_strongly_measurable_snd` should deduce that the second function is `ae_strongly_measurable` from the fact that the first one is. And there is also the statement that, in a second-countable space, both functions are automatically `ae_strongly_measurable` (as this follows from `ae_measurable`), which I had named `ae_strongly_measurable` (and not `ae_strongly_measurable_fst`, to avoid the ambiguous symmetry between two statements saying different things). That's why I had also added similar `ae_measurable` versions, for symmetry. But you are right that this is ambiguous, so I removed `ae_measurable` and `ae_strongly_measurable`, renaming the latter to `ae_strongly_measurable_fst` (and adding a docstring to warn the reader that it has nothing to do with `ae_strongly_measurable_snd`).